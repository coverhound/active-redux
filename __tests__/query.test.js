import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { Attr, bind, define } from 'active-redux';
import QueryProxy from 'active-redux/query';
import createMockStore from 'fixtures/store';

describe('QueryProxy', () => {
  let mockStore = createMockStore();

  beforeEach(() => {
    mockStore = createMockStore();
    bind(mockStore);
  });
  const mockAxios = new MockAdapter(axios);
  const blankPromise = new Promise(() => {});
  const httpHeaders = { 'content-type': 'application/json' };

  const Comment = define('comments', class Comment {});
  const Person = define('people', class Person {
    static attributes = { comments: Attr.hasMany('comments') }
  });
  const CommentProxy = new QueryProxy(Comment);
  const PersonProxy = new QueryProxy(Person);


  describe('#peek()', () => {
    it('returns the resource', () => {
      const result = CommentProxy.peek(5);
      expect(result).toMatchSnapshot();
    });
  });

  describe('#find()', () => {
    const remoteComment = Object.assign({}, mockStore.getState().api.resources.comments['5'], {
      attributes: { body: '[redacted]' },
    });

    test('it resolves to all resources in the store', async () => {
      mockAxios.onGet().replyOnce(200, { data: remoteComment }, httpHeaders);
      const results = await CommentProxy.find(5);
      expect(results).toMatchSnapshot();
    });
  });

  describe('#select()', () => {
    const remoteComment = Object.assign({}, mockStore.getState().api.resources.comments['5'], {
      attributes: { body: '[redacted]' },
    });

    test('while fetching if the resource is not in the store, it resolves to null', () => {
      mockAxios.onGet().replyOnce(() => blankPromise);
      const response = CommentProxy.select(1)(mockStore.getState());
      expect(response).toBe(null);
    });

    test('while fetching if the resource is in the store, it returns that resource', () => {
      mockAxios.onGet().replyOnce(() => blankPromise);
      const response = CommentProxy.select(5)(mockStore.getState());
      expect(response).toMatchSnapshot();
    });

    test('after fetching it returns the fetched resource', async () => {
      mockAxios.onGet().replyOnce(200, { data: remoteComment }, httpHeaders);
      const selector = CommentProxy.select(5);
      await new Promise((res) => setTimeout(res, 100));

      const response = selector(mockStore.getState());
      expect(response.data).toEqual(remoteComment);
    });
  });

  describe('#peekAll()', () => {
    it('returns the resources', () => {
      const response = PersonProxy.peekAll();
      expect(response).toMatchSnapshot();
    });
  });

  describe('#findAll()', () => {
    const remoteComments = [{
      type: 'comments',
      id: '132',
      attributes: { body: 'nice' },
    }, {
      type: 'comments',
      id: '215',
      attributes: { body: 'eh' },
    }];

    test('it resolves to the fetched resources', async () => {
      mockAxios.onGet().replyOnce(200, { data: remoteComments }, httpHeaders);
      const results = await CommentProxy.findAll();
      expect(results).toMatchSnapshot();
    });
  });

  describe('#selectAll()', () => {
    const remoteComments = [{
      type: 'comments',
      id: '132',
      attributes: { body: 'nice' },
    }, {
      type: 'comments',
      id: '215',
      attributes: { body: 'eh' },
    }];

    test('while fetching it returns what\'s in the store', () => {
      mockAxios.onGet().replyOnce(() => blankPromise);
      const response = CommentProxy.selectAll()(mockStore.getState());
      expect(response).toMatchSnapshot();
    });

    test('after fetching returns what\'s in the store', async () => {
      mockAxios.onGet().replyOnce(200, { data: remoteComments }, httpHeaders);
      const selector = CommentProxy.selectAll();
      await new Promise((res) => setTimeout(res, 100));

      const results = selector(mockStore.getState());
      expect(results).toMatchSnapshot();
    });
  });

  describe('#query()', () => {
    const remoteComments = [{
      type: 'comments',
      id: '140',
      attributes: { body: 'firetruck!' },
    }, {
      type: 'comments',
      id: '581',
      attributes: { body: 'fishing?' },
    }];
    const query = { page: 2, filter: { body: 'f*' } };
    const queryString = 'comments?page=2&filter%5Bbody%5D=f%2A';

    test('it resolves to the fetched resources', async () => {
      mockAxios.onGet(queryString).reply(200, { data: remoteComments }, httpHeaders);
      const results = await CommentProxy.query(query);
      expect(results).toMatchSnapshot();
    });
  });

  describe('#selectQuery()', () => {
    const remoteComments = [{
      type: 'comments',
      id: '140',
      attributes: { body: 'firetruck!' },
    }, {
      type: 'comments',
      id: '581',
      attributes: { body: 'fishing?' },
    }];
    const query = { page: 2, filter: { body: 'f*' } };
    const queryString = 'comments?page=2&filter%5Bbody%5D=f%2A';

    test('while fetching it returns what\'s in the store', () => {
      mockAxios.onGet(queryString).reply(() => blankPromise);
      const response = CommentProxy.selectQuery(query)(mockStore.getState());
      expect(response).toMatchSnapshot();
    });

    test('after fetching returns what\'s in the store', async () => {
      mockAxios.onGet(queryString).reply(200, { data: remoteComments }, httpHeaders);
      const selector = CommentProxy.selectQuery(query);
      await new Promise((res) => setTimeout(res, 100));

      const results = selector(mockStore.getState());
      expect(results).toMatchSnapshot();
      expect(results.isFetching).toBe(false);
    });
  });
});
