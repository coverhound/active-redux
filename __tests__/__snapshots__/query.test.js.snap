// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryProxy #find() it resolves to all resources in the store 1`] = `
Comment {
  "data": Object {
    "attributes": Object {
      "body": "[redacted]",
    },
    "id": "5",
    "links": Object {
      "self": "http://example.com/comments/5",
    },
    "relationships": Object {
      "author": Object {
        "data": Object {
          "id": "9",
          "type": "people",
        },
      },
    },
    "type": "comments",
  },
}
`;

exports[`QueryProxy #findAll() it resolves to the fetched resources 1`] = `
Array [
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "links": Object {
        "self": "http://example.com/comments/5",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "links": Object {
        "self": "http://example.com/comments/12",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "nice",
      },
      "id": "132",
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "eh",
      },
      "id": "215",
      "type": "comments",
    },
  },
]
`;

exports[`QueryProxy #peek() returns the resource 1`] = `
Comment {
  "data": Object {
    "attributes": Object {
      "body": "First!",
    },
    "id": "5",
    "links": Object {
      "self": "http://example.com/comments/5",
    },
    "relationships": Object {
      "author": Object {
        "data": Object {
          "id": "9",
          "type": "people",
        },
      },
    },
    "type": "comments",
  },
}
`;

exports[`QueryProxy #peekAll() returns the resources 1`] = `
Array [
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "123 Main St",
        },
        "first-name": "Dan",
        "last-name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "links": Object {
        "self": "http://example.com/people/9",
      },
      "relationships": Object {
        "comments": Object {
          "data": Array [
            Object {
              "id": "5",
              "type": "comments",
            },
            Object {
              "id": "12",
              "type": "comments",
            },
          ],
        },
      },
      "type": "people",
    },
  },
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "A different Road",
        },
        "first-name": "Bob",
        "last-name": "Dyland",
        "twitter": "shredmaster",
      },
      "id": "14",
      "links": Object {
        "self": "http://example.com/people/14",
      },
      "type": "people",
    },
  },
]
`;

exports[`QueryProxy #query() it resolves to the fetched resources 1`] = `
Array [
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "firetruck!",
      },
      "id": "140",
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "fishing?",
      },
      "id": "581",
      "type": "comments",
    },
  },
]
`;

exports[`QueryProxy #select() while fetching if the resource is in the store, it returns that resource 1`] = `
Comment {
  "data": Object {
    "attributes": Object {
      "body": "First!",
    },
    "id": "5",
    "links": Object {
      "self": "http://example.com/comments/5",
    },
    "relationships": Object {
      "author": Object {
        "data": Object {
          "id": "9",
          "type": "people",
        },
      },
    },
    "type": "comments",
  },
}
`;

exports[`QueryProxy #selectAll() after fetching returns what's in the store 1`] = `
Array [
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "links": Object {
        "self": "http://example.com/comments/5",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "links": Object {
        "self": "http://example.com/comments/12",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "nice",
      },
      "id": "132",
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "eh",
      },
      "id": "215",
      "type": "comments",
    },
  },
]
`;

exports[`QueryProxy #selectAll() while fetching it returns what's in the store 1`] = `
Array [
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "links": Object {
        "self": "http://example.com/comments/5",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "links": Object {
        "self": "http://example.com/comments/12",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
]
`;

exports[`QueryProxy #selectQuery() after fetching returns what's in the store 1`] = `
Array [
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "firetruck!",
      },
      "id": "140",
      "type": "comments",
    },
  },
  Comment {
    "data": Object {
      "attributes": Object {
        "body": "fishing?",
      },
      "id": "581",
      "type": "comments",
    },
  },
]
`;

exports[`QueryProxy #selectQuery() while fetching it returns what's in the store 1`] = `Array []`;
